pkg/gardenlet/controller/controllerinstallation/controllerinstallation/reconciler.go

```diff
@@ -469,3 +484,35 @@ func (r *Reconciler) reconcileGardenAccessSecret(ctx context.Context, controller
 
 	return accessSecret, accessSecret.Reconcile(ctx, r.SeedClientSet.Client())
 }
+
+// BootstrapControlPlaneNodeFunc adapts host network, replicas, tolerations and usable ports range for autonomous shoot
+// clusters if necessary.
+func (r *Reconciler) BootstrapControlPlaneNodeFunc(obj runtime.Object) error {
+	if !r.BootstrapControlPlaneNode {
+		return nil
+	}
+
+	if deployment, ok := obj.(*appsv1.Deployment); ok {
+		deployment.Spec.Replicas = ptr.To(int32(1))
+	}
+	return kubernetesutils.VisitPodSpec(obj, func(podSpec *corev1.PodSpec) {
+		podSpec.HostNetwork = r.BootstrapControlPlaneNode
+		podSpec.Tolerations = append(podSpec.Tolerations,
+			corev1.Toleration{Operator: corev1.TolerationOpExists, Effect: corev1.TaintEffectNoSchedule},
+			corev1.Toleration{Operator: corev1.TolerationOpExists, Effect: corev1.TaintEffectNoExecute},
+		)
+	})
+}
+
+// CalculateNextUsablePorts returns the next usable port range for the next controller installation.
+func (r *Reconciler) CalculateNextUsablePorts() []int {
+	if r.nextUsablePort == nil {
```

Comment:
Does `nextUsablePort` need to be a pointer? Could we use a zero value instead of the `nil` value?